src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs

```diff
@@ -96,12 +97,107 @@ public static void Fill(Span<byte> data)
+            if (leadingZeroCount == 0)
+            {
+                mask = unchecked((uint)-1);
+            }
+            else
+            {
+                mask = (1U << 32 - leadingZeroCount) - 1;
+            }
+
+            // We only want to generate as many bytes as required to satisfy the mask to not apply
+            // undue pressure to the underlying random number generator.
+            int bytesRequired = (32 + 7 - leadingZeroCount) / 8;
+
+            Span<uint> valueResult = stackalloc uint[1];
+            ref uint result = ref valueResult[0];
+            Span<byte> valueResultBytes = MemoryMarshal.AsBytes(valueResult);
+            if (BitConverter.IsLittleEndian)
+            {
+                do
+                {
+                    Span<byte> randomFillSlice = valueResultBytes.Slice(0, bytesRequired);
+                    RandomNumberGeneratorImplementation.FillSpan(randomFillSlice);
+                }
+                while ((result & mask) > maskedRange);
+            }
+            else
+            {
+                do
+                {
+                    Span<byte> randomFillSlice = valueResultBytes.Slice(bytesRequired);
+                    RandomNumberGeneratorImplementation.FillSpan(randomFillSlice);
+                }
+                while ((result & mask) > maskedRange);
+            }
+            return (int)((result & mask) + fromInclusive);
+        }
+
+        public static int GetInt32(int toExclusive)
+        {
+            if (toExclusive <= 0) throw new ArgumentOutOfRangeException(nameof(toExclusive), SR.ArgumentOutOfRange_NeedPosNum);
+            return GetInt32(0, toExclusive);
+        }
+
         internal void VerifyGetBytes(byte[] data, int offset, int count)
         {
             if (data == null) throw new ArgumentNullException(nameof(data));
             if (offset < 0) throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedNonNegNum);
             if (count < 0) throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNum);
             if (count > data.Length - offset) throw new ArgumentException(SR.Argument_InvalidOffLen);
         }
+
+        private static int LeadingZeroCount(uint value)
+        {
+            //TODO: Use lzcnt intrinsics when available.
+            int c = 1;
+            if ((value & 0xFFFF0000) == 0)
+            {
+                value <<= 16;
+                c += 16;
+            }
+            if ((value & 0xFF000000) == 0)
+            {
+                value <<= 8;
+                c += 8;
+            }
+            if ((value & 0xF0000000) == 0)
+            {
+                value <<= 4;
+                c += 4;
+            }
+            if ((value & 0xC0000000) == 0)
+            {
+                value <<= 2;
+                c += 2;
+            }
+            return c + ((int)value >> 31);
```

Comment:
I'm not sure this is properly implemented.

```
value = 00000000
c = 1
yes
value = 00000000
c = 17
yes
value = 00000000
c = 25
yes
value = 00000000
c = 29
yes
value = 00000000
c = 31

return 31 + (0 >> 31) => 31
(32 was correct)
```

The "maybe subtraction" at the end seems less straightforward than the clz3 algorithm on Wikipedia (https://en.wikipedia.org/wiki/Find_first_set#CLZ).  Of course, that one also has to special case 0 :smile:.

0 won't be used in context here, but since this is supposed to be functionally equivalent to lzcnt it should handle 0 correctly, particularly because this is likely to serve as the implementation on arm.

Reply:
Hm, OK. I borrowed it from here and eyeballed it and it looked OK, I will re-implement. https://github.com/dotnet/corefx/blob/fc169cddedb6820aaabbdb8b7bece2a3df0fd1a5/src/Common/src/CoreLib/System/Decimal.DecCalc.cs#L790